src/main.cpp

```diff
@@ -3096,18 +3396,50 @@ static bool ActivateBestChainStep(CValidationState& state, const CChainParams& c
         if (!DisconnectTip(state, chainparams.GetConsensus()))
             return false;
         fBlocksDisconnected = true;
+
+        // Once the first block has been disconnected and a re-org has begun then we need to terminate any
+        // currently running PV threads that are validating.  They will likley have self terminated
+        // at this point anyway because the chain tip and UTXO base view will have changed but just
+        // to be sure we are not waiting on script threads to finish we can issue the termination here.
```

Comment:
I put it after the first DisconnectTip() because if the disconnect fails it will return false and therefore we don't want to kill the validation threads if there are any running .  (although it seems rather unlikely that a disconnect would fail and if it did i think there would be a more serious problem there, not sure it's even necessary to put in the code that it should return, perhaps an assert would be more appropriate).

Also I put it there in that section because I thought it would be best to kill the PV threads as soon as the first block was disconnected ...there may be other blocks to disconnect in the the event of a larger re-org so rather than waiting until all the blocks are disconnected , the PV threads are terminated immediately after the chaintip they are working on has been undone.